### DISCUSSION TITLE
Values construct does not work with UUIDs

### DISCUSSION THREAD

**Discussion by atellier** [sqlalchemy/sqlalchemy#6268] [Usage Questions]
**Describe the bug**
When joining with `values()` construct with `UUID` (using the implementation provided in the docs), the actual values are not surrounded by single quotes when sent to PostgreSQL and when using `literal_binds=True`. We are using `literal_binds` becuase we also have MSSQL server that use the same code and parameters count is limited to 2,100 by default on MSSQL.

**Expected behavior**
`values()` should cast or surround the 'UUIDs' with strings

**To Reproduce**
```py
import uuid
from sqlalchemy import select, create_engine

from sqlalchemy import Column, Float, values, column
from sqlalchemy.dialects.mssql import UNIQUEIDENTIFIER
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import declarative_base, sessionmaker
from sqlalchemy.types import TypeDecorator, CHAR


class UUID(TypeDecorator):
    """Platform-independent GUID type.

    Uses PostgreSQL's UUID type, otherwise uses
    CHAR(32), storing as stringified hex values.

    """

    impl = CHAR

    def load_dialect_impl(self, dialect):
        if dialect.name == "postgresql":
            return dialect.type_descriptor(PGUUID())
        elif dialect.name == "mssql":
            return dialect.type_descriptor(UNIQUEIDENTIFIER())
        else:
            return dialect.type_descriptor(CHAR(32))

    def process_bind_param(self, value, dialect):
        if value is None:
            return value
        elif dialect.name == "postgresql" or dialect.name == "mssql":
            return str(value)
        else:
            if not isinstance(value, uuid.UUID):
                return "%.32x" % uuid.UUID(value).int
            else:
                # hexstring
                return "%.32x" % value.int

    def process_result_value(self, value, dialect):
        if value is None:
            return value
        else:
            if not isinstance(value, uuid.UUID):
                value = uuid.UUID(value)
            return value


Base = declarative_base()
metadata = Base.metadata

class Test(Base):
    __tablename__ = "Test"
    ID = Column(UUID, primary_key=True)
    Value = Column(Float, nullable=False)

engine = create_engine(...)
Base.metadata.create_all(
    engine,
    tables=[
        Test.__table__,
    ],
    checkfirst=True,
)

uuids = [uuid.uuid4() for _ in range(2)]

Session = sessionmaker(engine)
with Session() as session:
    session.add_all([Test(ID=uuid, Value=1) for uuid in uuids])
    session.commit()

    filter_query = select(
        values(
            column("ID", Test.ID.type),
            name="TestID",
            literal_binds=True,
        ).data([(_id,) for _id in uuids])
    ).subquery()
    session.execute(
        select(Test).join(filter_query, filter_query.c.ID == "ID")
    ).scalars().all()
```

**Error**

```
sqlalchemy.exc.StatementError: (builtins.ValueError) badly formed hexadecimal UUID string
[SQL: SELECT "Test"."ID", "Test"."Value" 
FROM "Test" JOIN (SELECT "TestID"."ID" AS "ID" 
FROM (VALUES (8211b998-1d94-4798-bf92-f3c8f976131b), (8bd9699e-6f72-49ef-9036-a1d0b4bba4ab)) AS "TestID" ("ID")) AS anon_1 ON anon_1."ID" = %(ID_1)s]
[parameters: [{}]]
```

**Versions.**
 - OS: Windows 10
 - Python: 3.8.8
 - SQLAlchemy: 1.4.7
 - Database: PostgreSQL 12
 - DBAPI: psycopg2

**Additional context**
interestingly, the query object seems to have the single quotes around it.  The query may not work since there is no cast to a UUID, but at least the string is quoted - see attached screen

Thanks
![Snag_5a65a1d](https://user-images.githubusercontent.com/889679/114759506-dea98580-9d23-11eb-88ff-c46b1134e027.png)

**Comment by CaselIT** [comment]
Hi,

The output is like that since it's executing what you have specified in `process_bind_param`:
```
elif dialect.name == "postgresql" or dialect.name == "mssql":
            return str(value)
```

SQLAlchemy is delegating the conversion to literal values to your implementation, to it has no way of knowing how to change what is returned. If you remove `literal_binds` or adjust the `process_bind_param` to return the correct value it should work.

**Comment by atellier** [comment]
using the following to add quotes around the value for instance: 
```
        elif dialect.name == "postgresql" or dialect.name == "mssql":
            return str(f"'{value}'")
```
then breaks the insert:
```
sqlalchemy.exc.StatementError: (builtins.ValueError) badly formed hexadecimal UUID string
[SQL: INSERT INTO "Test" ("ID", "Value") VALUES (%(ID)s, %(Value)s)]
[parameters: [{'ID': UUID('f0a4e27a-dfc1-452f-833f-78583101e460'), 'Value': 1}, {'ID': UUID('39d34cb1-b3fc-4d82-9380-1d2b02d5cf53'), 'Value': 1}]]
```
So this is why I suspected the issue could have been with `values` instead. May be `values` should cast to the actual type? 

I ended having to do a code like that to specifically catch the case when I have a UUID:
```
    if isinstance(Test.ID.type, UUID):
        filter_query = select(
            values(
                column("ID", String),
                name="TestID",
                # MSSQL does not support more than 2,100 parameters
                literal_binds=True,
            ).data([(str(_id),) for _id in uuids])
        ).subquery()
        session.execute(
            select(Test).join(filter_query, cast(filter_query.c.ID, UUID) =="ID")
        ).scalars().all()
```
and manually cast to UUID from a string, otherwise PostgreSQL would complain on mismatch between varchar and UUID.

**Reply by CaselIT** [reply]
I would suggest looking at the postgres types for hints
https://github.com/sqlalchemy/sqlalchemy/blob/20c0f774e5517514da811bc446812baa6b1f32f1/lib/sqlalchemy/dialects/postgresql/base.py#L1678-L1734
and
https://github.com/sqlalchemy/sqlalchemy/blob/20c0f774e5517514da811bc446812baa6b1f32f1/lib/sqlalchemy/dialects/postgresql/pygresql.py#L144-L161

> So this is why I suspected the issue could have been with values instead. May be values should cast to the actual type?

Not sure if that can be generalized. I believe that you can cast the lindparms values by returning ` return f"'{value}'::uuid"` in pg
